code/data_structures/tree/binary_search_tree/binary_search_tree.cpp

```diff
@@ -1,88 +1,377 @@
-#include<iostream>
-using namespace std;
+/*
+** The C++ Implementation of Binary Search Tree
+**
+** Binary Search Tree, is a node-based binary tree data 
+** structure which has the following properties:
+**
+** - The left subtree of a node contains only nodes with 
+** keys less than the node’s key.
+** - The right subtree of a node contains only nodes with 
+** keys greater than the node’s key.
+** - The left and right subtree each must also be a binary 
+** search tree.
+*/
 
-struct node
+#include <iostream>
+
+using std::cout;
+using std::endl;
+
+struct BinaryTreeNode
 {
-    int key;
-    struct node *left, *right;
+	int _value;
+
+	BinaryTreeNode *_Left;
+	BinaryTreeNode *_Right;
+
+	BinaryTreeNode() : _value(0), _Left(nullptr), _Right(nullptr)
+	{	}
+
+	BinaryTreeNode(int value) : _value(value), _Left(nullptr), _Right(nullptr)
+	{	}
+};
+
+class BST
+{
+public:
+	BST() : _Root(nullptr), _count(0) {	}
+	~BST() { Clear(_Root); }
+
+	void Add(int);
+	void Remove(int);
+	int FindSmallest(void);
+	void Clear(BinaryTreeNode *);
+	bool Contains(int);
+	int Count(void) { return _count; }
+
+	BinaryTreeNode * Root(void) { return _Root; }
+
+	void preOrder(BinaryTreeNode *);
+	void postOrder(BinaryTreeNode *);
+	void inOrder(BinaryTreeNode *);
+
+private:
+	int _count;
```

Comment:
For all of your private member variables, suffix then with an underscore, not prefix them. The names starting with an underscore are reserved by the compiler.